libraries/fbx/src/GLTFSerializer.cpp

```diff
@@ -1536,81 +1536,84 @@ bool GLTFSerializer::buildGeometry(HFMModel& hfmModel, const hifi::VariantHash&
 
-                    for (int weightedIndex = 0; weightedIndex < values.size(); ++weightedIndex) {
-                        float weight = 0.1f;
+                    for (int weightedIndex = 0; weightedIndex < keys.size(); ++weightedIndex) {
+                        float weight = 1.0f;
                         int indexFromMapping = weightedIndex;
                         int targetIndex = weightedIndex;
                         hfmModel.blendshapeChannelNames.push_back("target_" + QString::number(weightedIndex));
 
-                        if (!names.isEmpty()) {
+                        if (!names.isEmpty() && names.contains(keys[weightedIndex])) {
                             targetIndex = names.indexOf(keys[weightedIndex]);
                             indexFromMapping = values[weightedIndex].first;
-                            weight = weight * values[weightedIndex].second;
+                            weight = values[weightedIndex].second;
                             hfmModel.blendshapeChannelNames[weightedIndex] = keys[weightedIndex];
                         }
                         HFMBlendshape& blendshape = mesh.blendshapes[indexFromMapping];
-                        blendshape.indices = part.triangleIndices;
                         auto target = primitive.targets[targetIndex];
 
                         QVector<glm::vec3> normals;
                         QVector<glm::vec3> vertices;
 
-                        if (weights.size() == primitive.targets.size()) {
-                            int targetWeight = weights[targetIndex];
-                            if (targetWeight != 0) {
-                                weight = weight * targetWeight;
-                            }
-                        }
-
                         if (target.values.contains((QString) "NORMAL")) {
                             generateTargetData(target.values.value((QString) "NORMAL"), weight, normals);
                         }
                         if (target.values.contains((QString) "POSITION")) {
                             generateTargetData(target.values.value((QString) "POSITION"), weight, vertices);
                         }
-                        bool isNewBlendshape = blendshape.vertices.size() < vertices.size();
-                        int count = 0;
-                        for (int i : blendshape.indices) {
-                            if (isNewBlendshape) {
-                                blendshape.vertices.push_back(vertices[i]);
-                                blendshape.normals.push_back(normals[i]);
-                            } else {
-                                blendshape.vertices[count] = blendshape.vertices[count] + vertices[i];
-                                blendshape.normals[count] = blendshape.normals[count] + normals[i];
-                                ++count;
-                            }
+
+                        if (blendshape.indices.size() < prevMeshVerticesCount + vertices.size()) {
```

Comment:
why do we need to do this?  we resize it but then only set the things at the end?

Reply:
This is handling the case that the glTF mesh comprises two glTF primitives. The primitives are iterated over in the loop...
```
foreach(auto &primitive, _file.meshes[node.mesh].primitives) {
```
Primitive vertices are pushed onto the the back of `mesh.vertices`. We need to set the blendshape data at the same indexes.

Reply:
right, I guess my question is, can we add prevMeshVerticesCount to the blendshape index without enlarging the blendshape itself?

e.g. remove the resizes + the other uses of prevMeshVerticesCount, except when we set the index to prevMeshVerticesCount + i

Reply:
Say, a mesh has a blendshape. (Blendshapes are specified at the mesh level.)
Say, the mesh has two primitives (submeshes). Both must target the blendshape.
Primitive 0 is processed. It has 10 vertices, say. So 10 vertices are added to `mesh.vertices` at indexes `0 .. 9` and 10 blendshape target values are added to each of the relevant `blendshape.indices`, `blendshape.vertices`, and `blendshape.normals`, also at indices `0 - 9`.
Primitive 1 is processed. It has 5 vertices, say. So 5 further items are added to each of `mesh.vertices` and the relevant `blendshape.indices` etc., at indices `10 - 14`.

It's not possible to get a primitive (submesh) with a blendshape without preceding primitives also having the same blendshape. And, besides, the `hfm::Model` class doesn't have any facility for specifying the start index for blendshape target data.

We could do away with explicit `resize()` statements and use either `push_back()` to add a new blendshape target value or use `[ ] ` to update an existing blendshape target value, depending on whether or not the blendshape already has a value at the target index. But this seems more complex logic, to me.

Reply:
ok, I *think* I understand.  I'm a little confused because I don't see similar code for FBX, and the old code here didn't seem to do this, but it sounds like you're effectively adding support for "blendshapes on meshes with multiple submeshes" here, and since blendshapes are specified per mesh, it's not like we'll ever unnecessarily resize these to say 15 and then only set the last 5 values; the first 10 values will have already been set from the first submesh pass

Reply:
This is correct. Except that in glTF a mesh is defined as one or more "primitives" (which is where the vertices, for example, reside). The old code only supported blendshapes on a mesh having one primitive (and even then it didn't work properly). It's not really "submeshes" in the FBX sense.